Dockerfile

```diff
@@ -0,0 +1,41 @@
+# https://docs.docker.com/engine/reference/builder/#from
+#   "The FROM instruction initializes a new build stage and sets the
+#    Base Image for subsequent instructions."
+FROM erlang:20.3.8.1-alpine as builder
+# https://docs.docker.com/engine/reference/builder/#label
+#   "The LABEL instruction adds metadata to an image."
+LABEL stage=builder
+
+# Install git for fetching non-hex depenencies. Also allows rebar3
+# to find it's own git version.
+# Add any other Alpine libraries needed to compile the project here.
+# See https://wiki.alpinelinux.org/wiki/Local_APK_cache for details
+# on the local cache and need for the symlink
+RUN ln -s /var/cache/apk /etc/apk/cache && \
+    apk update && \
+    apk add --update openssh-client git 
+
+# WORKDIR is located in the image
+#   https://docs.docker.com/engine/reference/builder/#workdir
+WORKDIR /root/rebar3
+
+# copy the entire src over and build
+COPY . .
+RUN ./bootstrap
+
+# this is the final runner layer, notice how it diverges from the original erlang
+# alpine layer, this means this layer won't have any of the other stuff that was
+# generated previously (deps, build, etc)
+FROM erlang:20.3.8.1-alpine as runner
+
+# copy the generated `rebar3` binary over here
+COPY --from=builder /root/rebar3/_build/prod/bin/rebar3 .
+
+# and install it
+RUN install -v ./rebar3 /usr/local/bin/ \
```

Comment:
I'd prefer this used `rebar3 local install`. You can set `HOME` to `/opt` in the instruction and then create a link from `/usr/local/bin/rebar3` to `/opt/.cache/rebar3/bin/rebar3`.

Patch I've been meaning to make is to have `local install` properly use `XDG_CACHE_DIR` and then you could set that to `/opt/` and it'd install to `/opt/rebar3/` instead of that `.cache` directory.

Reply:
updated

Reply:
update the comment above. With these changes, you'll also need to copy over `/opt/.cache/rebar3` besides the `rebar3` binary, is this better somehow?

Reply:
Oh, didn't think about the use of copying it into another image. Argh, annoying, it is far better to use the unpacked version when calling from Entrypoint, so people default to that, but is a pain when copying.

I guess you can just undo my suggested change. It'd probably be confusing if there was both the rebar3 escript and the local install copy.

Reply:
If properly documented, people will copy/paste the two lines over to their own docker images. What is being offered through the `local install` method additionally to `install -v`?
btw, `install -v` is the method used in the [official Erlang docker images](https://github.com/erlang/docker-erlang-otp/blob/master/23/alpine/Dockerfile)

Reply:
Yea, I want to change that in those too.

local install offers it not being an escript, which among making it slightly more efficient means the terminal doesn't get all screwed up if the user `ctrl-c`'s out of the shell.

Reply:
ah nice, we should keep `local install` then